core-bundle/src/Security/TwoFactor/BackupCode/BackupCodeManager.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\TwoFactor\BackupCode;
+
+use Contao\User;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Backup\BackupCodeManagerInterface;
+
+class BackupCodeManager implements BackupCodeManagerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function isBackupCode($user, string $code): bool
+    {
+        if (!$user instanceof User) {
+            return false;
+        }
+
+        return $user->isBackupCode($code);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function invalidateBackupCode($user, string $code): void
+    {
+        if (!$user instanceof User) {
+            return;
+        }
+
+        $user->invalidateBackupCode($code);
+        $user->save();
+    }
+
+    public function generateBackupCodes(User $user): ?array
+    {
+        $backupCodes = [];
+
+        for ($i = 0; $i < 10; ++$i) {
+            $backupCodes[] = $this->generateCode();
+        }
+
+        $user->backupCodes = json_encode($backupCodes);
+        $user->save();
+
+        return $backupCodes;
+    }
+
+    private function generateCode(): string
+    {
+        return sprintf(
+            '%s-%s',
+            substr(uniqid(bin2hex(random_bytes(128)), true), 0, 5),
+            substr(uniqid(bin2hex(random_bytes(128)), true), 0, 5)
```

Comment:
Why this combination of functions with these parameters?

Reply:
Usually the backup codes are opaque random credentials, not mathematically based on the 2FA seed or anything else known about the user, and generated with sufficient length and entropy as to avoid dictionary attacks.

Reply:
But due to ``substr`` only the first 5 characters are relevant?

One byte equals two hex characters, therefore I assume ``substr(bin2hex(random_bytes(3)), 0, 5)`` is equally secure as ``substr(uniqid(bin2hex(random_bytes(128)), true), 0, 5)``?

Reply:
I think so, too. We can even omit the `substr()` IMHO and use `bin2hex(random_bytes(3))`. This will create a random 6 character hex string, so the overall backup code would be 13 characters long then.

Reply:
@bytehead I have changed this in 4511aba6b14c98ed2e3f94e6f86782318928e3ef. Feel free to revert the commit if you have concerns.